include/pktbuf.h

```diff
@@ -158,4 +180,16 @@ void pktbuf_write_ExtQuery(PktBuf *buf, const char *query, int nargs, ...);
 #define SEND_SASLResponseMessage(res, sk, cr) \
 	SEND_wrap(512, pkgbuf_write_SASLResponseMessage, res, sk, cr)
 
+#define SEND_ParseComplete(res, sk) \
+	SEND_wrap(5, pktbuf_write_ParseComplete, res, sk)
```

Comment:
I think we should do something similar when sending packets to clients, as we do when sending packets to the server. We should not use `pktbuf_send_immediate`, but instead something similar to `sbuf_queue_packet`. Otherwise these packets might be sent right in the middle of some other packet that we're forwarding from the server to the client. As discussed in another thread, keeping track of where exactly in the queue this packet is sent is overkill. But we should at least make sure it's not sent in the middle of some other packet. i.e. out of order packets seem acceptable, but corrupted packets are not acceptable.

Reply:
To be clear, I don't think `sbuf_queue_packet` works right now for these. As it's intended to also be able to send just the first part of the packet (to support the Bind optimization where we don't want to buffer all the arguments). So any packets queued with that function currently, would still be sent as soon as possible. We'd want something similar, but slightly different or we'd have to rethink `sbuf_queue_packet` a bit. I'll think a bit more about this at least.

Reply:
I wonder in which cases we need to send more than one response to the client?
It seems to be only possible if pipeline mode is used (https://www.postgresql.org/docs/current/libpq-pipeline-mode.html)
But in this case problems with `pktbuf_send_immediate` can happen not only with prepared statements: all other `SEND_*` macros are also using `SEND_wrap`.

Also I do not completely understand why `sbuf_queue_packet` doesn't work in this case? 

In any case - are you going to fix it yourself or should I try to do something?

Reply:
I will fix it myself

Reply:
BTW, the critical difference between existing usages of `pktbuf_send_immediate` and the new ones introduced in this PR is that these ones are used while a server is linked to a client. When that's the case it's not really safe to call `pktbuf_send_immediate` and `pktbuf_send_queued`.

I tried reproducing using pipeline mode of psycopg to trigger this to happen but I was unable to. I did however run into this issue when doing so: https://github.com/psycopg/psycopg/issues/589

Reply:
> keeping track of where exactly in the queue this packet is sent is overkill. 

After looking closer at what libpq was doing to implement pipelining, it turned out that keeping such a queue is exactly what libpq does for a subset of commands: https://github.com/postgres/postgres/blob/876d17d62fe84d706058f2458f067f0e9049a230/src/interfaces/libpq/libpq-int.h#L447-L452

I was able to create a pipelining test that would completely freeze the clientside, because it continued to wait for some message that PgBouncer sent out of order. So I spent some time actually implementing a command queue similar to the one in libpq, which made it work:  [`6335e76` (#845)](https://github.com/pgbouncer/pgbouncer/pull/845/commits/6335e76167a6d65ada1c33f04e33c2ee74bce4ab)

Reply:
This new code can definitely use some reviews btw. There's probably some issues with it. I added comments to different parts of the code that probably need some more changes.